### ISSUE TITLE
Memory Corruption w/ resize layer

### ISSUE THREAD

**Issue by nickcano** [gnawice/mojo-cnn#14]
I'm working with a network where the input layer isn't square, so I've used a resize layer to square it. Full topology (ignore the syntax, it's some basic wrappers which are unrelated to the issue):

```
addLayer(network, "I1",  "input 60 16 3")
addLayer(network, "RS1", "resize 30 zero")
addLayer(network, "CV1", "convolution 6 40 1 elu")
addLayer(network, "PL1", "semi_stochastic_pool 2 2")
addLayer(network, "CC1", "concatenate 15")
addLayer(network, "DP1", "deepcnet 50 elu")
addLayer(network, "DO1", "dropout 0.300000")
addLayer(network, "DP2", "deepcnet 100 elu")
addLayer(network, "DO2", "dropout 0.300000")
addLayer(network, "O1",  "softmax 6")
connectAll(network)
```

This trains fine on a single thread, but I get heap corruption if I:
1. try to multi-thread training
2. save the trained model, load it, and then destruct it after running some classifications

Now, as with most heap corruption, the area that throws the exception is rarely related to the actual issue; more often a side effect of the corrupted memory. To track down the source of the corruption, I fired up Application Verifier. With the heap checks enabled, app verifier aligns all allocations so there's a guarded page following them. This allows an exception to be thrown as soon as heap corruption happens, as any over-reads or -writes will hit the guarded page.

I found it to be happening [here](https://github.com/gnawice/mojo-cnn/blob/master/mojo/layer.h#L1698). The line in question is the first `memcpy()` call in `accumulate_signal()`:

```cpp
memcpy(node.x + node.chan_stride*map_offset, m.x, sizeof(float)*m.size());
```
According to my debugger, it happens because `node` is `30`x`30`x`3` (total size 2700) but `m` is `60`x`30`x`3` (total size 5400). This seems to be because the code has decided to call `top.node.pad()` for `top.node` of `60`x`16`x`3` because of the need for padding in the `y` dimension. It has failed to consider, however, the possibility that one dimension needs a crop while the other needs a pad.

Is there an intention to support this? Is there a better way to accomplish it (other than a pre-processing step)?

Also, one final question, what does `crop()` do? Is it a basic crop that will lose anything outside? Or is there some sort of stepping/squishing/interpolation going on? Due to not knowing the code base well, I'm having trouble completely understanding the function.

**Comment by nickcano** [comment]
Oh, it looks like the cropping in that function shouldn't even work:

```cpp
int dx = abs(padx) / 2;
int dy = abs(pady) / 2;
matrix m = top.node.crop(dx, dy, node.cols, node.rows, _thread_count);
memcpy(node.x + node.chan_stride*map_offset, m.x, sizeof(float)*m.size());
```
Because `padx` and `pady` will remain as `0` unless `opadx` or `opady` respectively are `> 0`:

```cpp
int padx=0, pady=0, padx_ex=0, pady_ex=0;

if (opadx > 0) padx = opadx/2;
if (opady > 0) pady = opady/2;
```

But the logic will only fall through to crop if both are `< 0` due to the first `if`:

```cpp
if (padx+ padx_ex > 0 || pady+ pady_ex > 0 )
```

As `padx+ padx_ex` and `pady+ pady_ex` are directly dependant on `opadx` and `opady`.

So, whatever the crop is trying to do with `dx` and `dy` will not be done, as they will always be `0`. Was the intention to do:

```cpp
int dx = abs(opadx) / 2;
int dy = abs(opady) / 2;
matrix m = top.node.crop(dx, dy, node.cols, node.rows, _thread_count);
memcpy(node.x + node.chan_stride*map_offset, m.x, sizeof(float)*m.size());
```

instead?

**Comment by nickcano** [comment]
Another case of memory corruption:

```cpp
	inline matrix& operator+=(const matrix &m2){
	  for(int i = 0; i < _size; i++) x[i] += m2.x[i];
	  return *this;
	}
```
Here, debugger says `this->_size` is `1600`, `m2.size` is `900`.

The call originates from `dropout_layer::distribute_delta()`, with `top` as my `DP2` layer and `this` as my `DO2` layer. Debugger says both of these are `3`x`3`x`100`, but `DP2` has size `1600` and `DO2` has size `900`. Why the mismatch?

**Comment by nickcano** [comment]
I updated the operator to do a proper bounding check:

```cpp
	inline matrix& operator+=(const matrix &m2) {
		int maxs = (_size > m2.size()) ? m2.size() : _size;
		for(int i = 0; i < maxs; i++)
			x[i] += m2.x[i];
		return *this;
	}
```

I did the same to all other `matrix` operators. Will see where that gets me.

**Comment by nickcano** [comment]
I also updated the block in `concatenation_layer::accumulate_signal()` to do what I _think_ may be proper padding and cropping. This replaces everything after the `int map_offset = layer_to_channel[&top];` line:

```cpp
		if ((node.cols == top.node.cols) && (node.rows == top.node.rows))
		{
			memcpy(node.x + node.chan_stride*map_offset, top.node.x, sizeof(float)*top.node.size());
		}
		else
		{
			matrix m = top.node;
			if (padx + padx_ex > 0 || pady + pady_ex > 0)
				m = m.pad(padx, pady, padx + padx_ex, pady + pady_ex, _pad_type, _thread_count);
			if (opadx < 0)
				m = m.crop(abs(opadx) / 2, 0, node.cols, node.rows, _thread_count);
			if (opady < 0)
				m = m.crop(0, abs(opady) / 2, node.cols, node.rows, _thread_count);
			memcpy(node.x + node.chan_stride*map_offset, m.x, sizeof(float)*m.size());
		}
```
There's no memory corruption, but I don't know how to check that the values are as they should be.

**Comment by nickcano** [comment]
I had a model at 100% (easy but important dataset) before I started running into all the corruption issues. Going to retrain with my changes and try to hit the same accuracy. I hope it works, but given how literred the code seems to be with overflows, I'm worried that some of the success may have been due to overreads succeeding and allowing processing of parts of the images that weren't supposed to be there with proper cropping, summing, etc. We'll see; epoch 8 and no corruptions yet.

**Comment by nickcano** [comment]
I really do appreciate this library. I've been needing something portable without a ton of dependencies, CPU optimized, powerful, and written in C++. Your library is all of those things. Thank you.

However, I feel the need to point out how *bad* these types of bugs can be. Yeah, they cause headaches and can steer people away from using your otherwise marvelous code, but, worse, they are security issues. In a worst case scenario, an attacker could attack one of these overwrites *from the input buffer*. It's not fun to imagine a machine getting hijacked while processing an image because the NN library was vulnerable to security flaws.

**Comment by nickcano** [comment]
Convolution layers with a stride of 2 appear to be overreading. Full topology:

```lua
addLayer(network, "I1",  "input 60 16 3")
addLayer(network, "RS1", "resize 38 zero")
addLayer(network, "CV1", "convolution 5 40 2 elu")
addLayer(network, "PL1", "semi_stochastic_pool 2 2")
addLayer(network, "CC1", "concatenate 20")
addLayer(network, "DP1", "deepcnet 25 elu")
addLayer(network, "DO1", "dropout 0.300000")
addLayer(network, "DP2", "deepcnet 50 elu")
addLayer(network, "DO2", "dropout 0.300000")
addLayer(network, "DP3", "deepcnet 100 elu")
addLayer(network, "DO3", "dropout 0.300000")
addLayer(network, "O1",  "softmax 6")
connectAll(network)
```
Inside of the loop in `convolution_layer::distribute_delta()`'s final `else` block for `CV1`, there is this call:

```cpp
for(int map=start_map; map<stop_map; map++) // how many maps  maps= node.chans
{
	top.delta.x[td_i] += unwrap_2d_dot_rot180(
		&delta_pad.x[i+(j)*delta_pad.cols + map*map_stride], 
		&w.x[(map+k*maps)*kernel_size],
		kernel_cols,
		delta_pad.cols,kernel_cols);

} // all input chans
```
The overread eventually happens in `delta_pad.x`. At this point, debugger says `i+(j)*delta_pad.cols + map*map_stride` is `18 + 32*25 + 39 * 625 = 25193`. It also says `delta_pad->_size` is `25000`, so that is enough to overread upon first usage already. It got into `unwrap_2d_dot_rot180()` and made it to `j = 4` for this line (it would have happened on `j = 0` if app verifier was on):

`v+= dot_rot180(&x1[stride1*j],&x2[stride2*(size-j-1)],size);`
With `stride1` as `25` and thus another `100` entries into the already overread array.

**I don't know how to fix this one**, I don't want to break CNNs, so I'm just gonna go back to stride of 1.

**Comment by gnawice** [comment]
Strides were partially supported but they do not work in the convolution layers. 

With respect to resizing the input.. you'll need to use a square input as only square inputs are supported currently. The resize wasn't tested for this use.  Resize was primarily put in to allow concat of layers.

**Comment by nickcano** [comment]
@gnawice your reply servery misses the point, and comes off as both rude and worrisome. I didn't write this bug report necessarily asking for help, but rather disclosing to you that your code has severe instability issues which are very very likely to be exploitable security holes.

If you read through my comments, you will find I've provided various fixes that make everything far more stable and secure. You'll find that I've outlined my process to help you understand how to easily track future issues down. However, since I do not fully understand the code, I am unsure if my fixes preserve the desired functionality. For example, in cases where I fixed buffer overflows, it's possible the real issue could have been an under-contained buffer. That's something to take note of if you apply these fixes.

Anyways, whether or not everything still works as intended, my fixes allowed me to train multiple models successfully, even without square inputs.

Regardless of that, it is a bad idea to release a library with these sorts of issues to then reply to users with `hey if you don't wanna get hacked make your input a square`. As I said before, I really like your library. But it is lacking things which would greatly improve stability, prevent security issues, and, seemingly to your pleasure, make users less likely to report issues.

Runtime assertions checking ranges and sizes, layer connections validating proper shape and connectivity, and some error log messages that make it clear when and why things are broken would be a great start.

It's clear that you're an expert in machine learning, and I appreciate that. I'm just trying to bring my expertise in security and development to the table.

**Comment by gnawice** [comment]
Whoa.. truly sorry if that came off as being rude. Certainly not my intention. Just trying to give a quick response because I didn't want you to think I was ignoring all your effort (which I appreciate). I don't have bandwidth right now to dig through all the issues. The short of it is that I'm not surprised there are issues if the input is not square. Similarly the stride is currently broken.  Just no time to fix it right now. 

The library has been useful for me because before opencv had a DNN module, there was basically no easy way to deploy CNNs. So this was made to be a minimal approach and error checking for some things was intentionally left out.  I don't think error checking is a big issue as you'll get errors when trying to get a network to run (as apposed to deployment).  I think the lack of documentation is the big issue as is absence of logging as it is not always clear when things don't work. Nevertheless, when i used mxnet and tensorflow, I similarly get errors because things don't work the way I wired them up. The only difference in that case is that I have stackoverflow to help.  

I do appreciate your feedback.  The next time I spend a weekend on the code, I'll address your points.
